libraries/AC_AttitudeControl/AC_PosControl.cpp

```diff
@@ -546,125 +593,67 @@ void AC_PosControl::calc_leash_length_z()
-        }
-    } else {
-        _accel_desired.z = 0.0f;
-    }
-
-    // store this iteration's velocities for the next iteration
-    _vel_last.z = _vel_target.z;
-
-    // reset velocity error and filter if this controller has just been engaged
-    if (_flags.reset_rate_to_accel_z) {
-        // Reset Filter
-        _vel_error.z = 0;
-        _vel_error_filter.reset(0);
-        _flags.reset_rate_to_accel_z = false;
-    } else {
-        // calculate rate error and filter with cut off frequency of 2 Hz
-        _vel_error.z = _vel_error_filter.apply(_vel_target.z - curr_vel.z, _dt);
-    }
-
-    _accel_target.z = _p_vel_z.get_p(_vel_error.z);
-
+    _accel_target.z = _pid_vel_z.update_all(_vel_target.z, curr_vel.z);
     _accel_target.z += _accel_desired.z;
 
-
+    // Acceleration Controller
 
     // Calculate Earth Frame Z acceleration
     const float z_accel_meas = get_z_accel_cmss();
 
-
     // ensure imax is always large enough to overpower hover throttle
     if (_motors.get_throttle_hover() * 1000.0f > _pid_accel_z.imax()) {
         _pid_accel_z.imax(_motors.get_throttle_hover() * 1000.0f);
     }
-
     float thr_out;
     if (_vibe_comp_enabled) {
-        _flags.freeze_ff_z = true;
-        _accel_desired.z = 0.0f;
-        const float thr_per_accelz_cmss = _motors.get_throttle_hover() / (GRAVITY_MSS * 100.0f);
-        // during vibration compensation use feed forward with manually calculated gain
-        // ToDo: clear pid_info P, I and D terms for logging
-        if (!(_motors.limit.throttle_lower || _motors.limit.throttle_upper) || ((is_positive(_pid_accel_z.get_i()) && is_negative(_vel_error.z)) || (is_negative(_pid_accel_z.get_i()) && is_positive(_vel_error.z)))) {
-            _pid_accel_z.set_integrator(_pid_accel_z.get_i() + _dt * thr_per_accelz_cmss * 1000.0f * _vel_error.z * _p_vel_z.kP() * POSCONTROL_VIBE_COMP_I_GAIN);
-        }
-        thr_out = POSCONTROL_VIBE_COMP_P_GAIN * thr_per_accelz_cmss * _accel_target.z + _pid_accel_z.get_i() * 0.001f;
+        thr_out = get_throttle_with_vibration_override();
     } else {
         thr_out = _pid_accel_z.update_all(_accel_target.z, z_accel_meas, (_motors.limit.throttle_lower || _motors.limit.throttle_upper)) * 0.001f;
+        thr_out += _pid_accel_z.get_ff() * 0.001f;
     }
     thr_out += _motors.get_throttle_hover();
 
+    // Actuator commands
+
     // send throttle to attitude controller with angle boost
     _attitude_control.set_throttle_out(thr_out, true, POSCONTROL_THROTTLE_CUTOFF_FREQ);
 
+    // Check for vertical controller health
+
     // _speed_down_cms is checked to be non-zero when set
     float error_ratio = _vel_error.z/_speed_down_cms;
-
     _vel_z_control_ratio += _dt*0.1f*(0.5-error_ratio);
     _vel_z_control_ratio = constrain_float(_vel_z_control_ratio, 0.0f, 2.0f);
 }
 
+// get throttle using vibration resistant calculation (uses feed forward with manually calculated gain)
+float AC_PosControl::get_throttle_with_vibration_override()
```

Comment:
Yes but at this point I'm not sure breaking this PR up more is going to help reduce complexity enough that it will lead to additional review.  My guess is after your review, there will be testing but nobody will have the confidence to dive into it and provide further feedback.